### DISCUSSION TITLE
Merging statements

### DISCUSSION THREAD

**Discussion by tbmuc** [ImageMagick/ImageMagick#3280] [Help] [answered]
Hello,

I would like to mege these 4 statements into one to increase the speed. The temporary files t1,t2,t3 are not required. I tried this by myself but got a lot of errors at each step. Can somebody help me?

magick  convert %1-crop.jpg -gravity north -crop %IMGSX%:%IMGSY%+0+0 +repage -resize %IMGSX%x%IMGSY%! t1.jpg
magick composite -gravity Northwest -geometry +%IMGPX%+%IMGPY% t1.jpg  -size %BSIZE% xc:#ffffff %t2.jpg		  
magick composite %TEMPLATE% t2.jpg %t3.jpg		
magick -size %FS2% -font %FONT% -background None -fill #000000 caption:%2 -trim -resize %FS1% -gravity center -extent %FS1% 1-txt.png
magick composite   -gravity Northwest -geometry  %FPOS% 1-txt.png t3.jpg t4.jpg

This is what it does.
1) Correct aspect ratio and size of the user providet image
2) Create white background and place the image to the right position
3) Overlay it with the template PNG.
4) Create a text with the best fit into a given box
5) Overlay the text 

here are the parameters just to help understanding. Please don't remove them.
set IMGSX=2343
set IMGSY=3256
set IMGPX=4241
set IMGPY=1078
set FPOSX=1256
set FPOSY=3784
set FPOS=+1256+3784
set FS1=2072x544
set FS2=4144x1088
set BSIZE=7056x5304

![t4small](https://user-images.githubusercontent.com/48103248/108591994-d6f4e280-736b-11eb-9500-dd8211f5c970.jpg)

**Comment by snibgo** [comment]
You don't provide sample input files, so I can't do this for you.

What is %TEMPLATE%? Worse, what is %t2.jpg and %t3.jpg?

Ignoring those problems, here are some guidelines. After each change you make, re-run the test. If the result if different then you have made the wrong change.

1. Change "magick convert" to just "magick".

2. Change each "magick composite" to just "magick". This needs the order of the two input files to be swapped. For example, the second command will be:
```
magick -gravity Northwest -geometry +%IMGPX%+%IMGPY% -size %BSIZE% xc:#ffffff t1.jpg %t2.jpg
```

3. Merge the first two commands. The first command does stuff to make t1.jpg, and the second uses t1.jpg. To merge them, replace t1.jpg in the second command with all the stuff from the first command.

4. Repeat for the other commands.

**Comment by tbmuc** [comment]
> magick -gravity Northwest -geometry +%IMGPX%+%IMGPY% -size %BSIZE% xc:#ffffff t1.jpg %t2.jpg

This is where the fun begins. An error occurs that gravity hat no image,
 -gravity Northwest -geometry +%IMGPX%+%IMGPY%  is related to t1.tmp
How to place it correctly?

magick  -size %BSIZE% xc:#ffffff  -gravity Northwest -geometry +%IMGPX%+%IMGPY% t1.jpg %t2.jpg 

results in a total white image

**Comment by snibgo** [comment, marked as answer]
Magick commands should generally be: read the inputs, do operations, write the output. I forgot to mention that when converting "magick composite" to "magick", you also need to insert a "-composite" operation.
```
magick -size %BSIZE% xc:#ffffff t1.jpg -gravity Northwest -geometry +%IMGPX%+%IMGPY% -composite t2.jpg
```
